alicloud/resource_alicloud_cms_sitemonitor.go

```diff
@@ -0,0 +1,192 @@
+
+import (
+	"fmt"
+	"strconv"
+	"time"
+
+	"github.com/aliyun/alibaba-cloud-sdk-go/services/cms"
+	"github.com/hashicorp/terraform-plugin-sdk/helper/resource"
+	"github.com/hashicorp/terraform-plugin-sdk/helper/schema"
+	"github.com/hashicorp/terraform-plugin-sdk/helper/validation"
+
+	"github.com/terraform-providers/terraform-provider-alicloud/alicloud/connectivity"
+)
+
+func resourceAlicloudCmsSiteMonitor() *schema.Resource {
+	return &schema.Resource{
+		Create: resourceAlicloudCmsSiteMonitorCreate,
+		Read:   resourceAlicloudCmsSiteMonitorRead,
+		Update: resourceAlicloudCmsSiteMonitorUpdate,
+		Delete: resourceAlicloudCmsSiteMonitorDelete,
+		Importer: &schema.ResourceImporter{
+			State: schema.ImportStatePassthrough,
+		},
+
+		Schema: map[string]*schema.Schema{
+			"address": {
+				Type:     schema.TypeString,
+				Required: true,
+			},
+			"task_name": {
+				Type:     schema.TypeString,
+				Required: true,
+			},
+			"task_type": {
+				Type:     schema.TypeString,
+				Required: true,
+				ForceNew: true,
+				ValidateFunc: validation.StringInSlice([]string{SiteMonitorHTTP, SiteMonitorDNS, SiteMonitorFTP, SiteMonitorPOP3,
+					SiteMonitorPing, SiteMonitorSMTP, SiteMonitorTCP, SiteMonitorUDP}, false),
+			},
+			"alert_ids": {
+				Type:     schema.TypeList,
+				Optional: true,
+				Elem:     &schema.Schema{Type: schema.TypeString},
+				MaxItems: 1,
+			},
+			"interval": {
+				Type:         schema.TypeInt,
+				Optional:     true,
+				Default:      1,
+				ValidateFunc: validation.IntInSlice([]int{1, 5, 15}),
+			},
+			"options_json": {
+				Type:     schema.TypeString,
+				Optional: true,
+			},
+			"task_state": {
+				Type:     schema.TypeString,
+				Computed: true,
+			},
+			"create_time": {
+				Type:     schema.TypeString,
+				Computed: true,
+			},
+			"update_time": {
+				Type:     schema.TypeString,
+				Computed: true,
+			},
+		},
+	}
+}
+
+func resourceAlicloudCmsSiteMonitorCreate(d *schema.ResourceData, meta interface{}) error {
+	client := meta.(*connectivity.AliyunClient)
+
+	taskName := d.Get("task_name").(string)
+	request := cms.CreateCreateSiteMonitorRequest()
+	request.Address = d.Get("address").(string)
+	request.TaskName = taskName
+	request.TaskType = d.Get("task_type").(string)
+	request.Interval = strconv.Itoa(d.Get("interval").(int))
+	request.OptionsJson = d.Get("options_json").(string)
+	alertIds := d.Get("alert_ids").([]interface{})
+	alertId := getAlertId(alertIds)
+	if alertId != "" {
+		request.AlertIds = alertId
+	}
+
+	_, err := client.WithCmsClient(func(cmsClient *cms.Client) (interface{}, error) {
+		return cmsClient.CreateSiteMonitor(request)
+	})
+	if err != nil {
+		return WrapError(err)
+	}
+
+	d.SetId(taskName)
+	return resourceAlicloudCmsSiteMonitorRead(d, meta)
+}
+
+func resourceAlicloudCmsSiteMonitorRead(d *schema.ResourceData, meta interface{}) error {
+	client := meta.(*connectivity.AliyunClient)
+	cmsService := CmsService{client}
+
+	keyword := d.Get("task_name").(string)
```

Comment:
Please use `d.Id()` instead of keyword.

Reply:
done, refactores service method ´DescribeSiteMonitor´to accept both id and keyword. keyword option is needed in Create method to get ID of site alarm since its id is not returned by ´CreateSiteMonitor´.